### PR TITLE
[CDAP-3523] [CDAP-3522] Implement REST APIs for searching Metadata

### PR DESCRIPTION
**Pull request by henry-cask**
CDAP-3523 Add new endpoint of REST APIs for searching metadata using value and key=value to metadata handler.

CDAP-3522 Add IndexedTable based search to the Business Metadata dataset.

Add more descriptive name for internal IndexedTable for business metadata dataset.

### CODE REVIEW

**Comment by henry-cask** [comment]
Build:
http://builds.cask.co/browse/CDAP-DUT2812-1

File: cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataHttpHandler.java [hunk 1]
```diff
@@ -344,4 +349,17 @@ public void removeStreamTags(HttpRequest request, HttpResponder responder,
...
+
+  @GET
+  @Path("/namespaces/{namespace-id}/metadata/search")
+  public void searchMetadata(HttpRequest request, HttpResponder responder,
+                             @PathParam("namespace-id") String namespaceId,
+                             @QueryParam("query") String searchQuery,
+                             @QueryParam("target") MetadataSearchTargetType target) throws Exception {
+    Iterable<MetadataSearchResultRecord> results = metadataAdmin.searchMetadata(searchQuery, target);
+
+    responder.sendJson(HttpResponseStatus.OK, Sets.newHashSet(results));
```
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataHttpHandler.java]: Should we just return a ``Set<MetadataSearchResultRecord>`` from ``metadataAdmin.searchMetadata(searchQuery, target)``? I made the change for the other methods in https://github.com/caskdata/cdap/pull/3891

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/service/BusinessMetadataStore.java [hunk 2]
```diff
@@ -167,6 +169,34 @@ public Void apply(BusinessMdsIterable input) throws Exception {
...
+   */
+  public Iterable<BusinessMetadataRecord> searchMetadataOnType(final String searchQuery,
+                                                        final MetadataSearchTargetType type) {
+    return txnl.executeUnchecked(new TransactionExecutor.Function<BusinessMdsIterable,
+      Iterable<BusinessMetadataRecord>> () {
+      @Override
+      public Iterable<BusinessMetadataRecord> apply(BusinessMdsIterable input) throws Exception {
+        // Currently we support two types of search formats: value and key=value.
+        // Check for exiatance of '=' char to make sure we did search in the right indexed column.
+        if (searchQuery.contains("=")) {
```
Comment (bdmogal) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/service/BusinessMetadataStore.java]: will ``=`` conflict with query params in REST API? If yes, should we use ``:`` instead?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/BusinessMetadataDataset.java [hunk 3]
```diff
@@ -187,13 +189,56 @@ public void removeMetadata(Id.NamespacedId targetId, Predicate<String> filter) {
...
+   * @return The {@link Iterable} of {@link BusinessMetadataRecord} that fit the key value pair.
+   */
+  public List<BusinessMetadataRecord> findBusinessMetadataOnKeyValue(String keyValue, MetadataSearchTargetType type) {
+    return executeSearchOnColumns(BusinessMetadataDataset.KEYVALUE_COLUMN, keyValue, type);
+  }
+
+  // Helper method to execute IndexedTable search on target index column.
+  List<BusinessMetadataRecord> executeSearchOnColumns(String column, String searchValue,
+                                                      MetadataSearchTargetType type) {
+    Scanner scanner = indexedTable.readByIndex(Bytes.toBytes(column), Bytes.toBytes(searchValue));
```
Comment (bdmogal) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/BusinessMetadataDataset.java]: Should ``KEYVALUE_COLUMN`` and ``VALUE_COLUMN`` be bytes, instead of doing this conversion every single time?

File: cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataHttpHandler.java (hunk 1 above)
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataHttpHandler.java]: Why a Set anyway? The underlying call from the Dataset actually returns a List.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/BusinessMetadataDataset.java [hunk 4]
```diff
@@ -187,13 +189,56 @@ public void removeMetadata(Id.NamespacedId targetId, Predicate<String> filter) {
...
+    List<BusinessMetadataRecord> results = new LinkedList<>();
+    try {
```
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/BusinessMetadataDataset.java]: move try one line up. It's a better practice to have try right after the resource creation. You can either have the List moved to the top or have the return statement moves inside the try.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/service/BusinessMetadataStore.java (hunk 2 above)
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/service/BusinessMetadataStore.java]: There shouldn't be conflict since query params need to be URL encoded.

```
path?p=key%3Dvalue
```

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/BusinessMetadataDataset.java [hunk 5]
```diff
@@ -187,13 +189,56 @@ public void removeMetadata(Id.NamespacedId targetId, Predicate<String> filter) {
...
+   * @param keyValue The metadata value to be found.
+   * @param type The target type of objects to search from.
+   * @return The {@link Iterable} of {@link BusinessMetadataRecord} that fit the key value pair.
+   */
+  public List<BusinessMetadataRecord> findBusinessMetadataOnKeyValue(String keyValue, MetadataSearchTargetType type) {
+    return executeSearchOnColumns(BusinessMetadataDataset.KEYVALUE_COLUMN, keyValue, type);
+  }
+
+  // Helper method to execute IndexedTable search on target index column.
+  List<BusinessMetadataRecord> executeSearchOnColumns(String column, String searchValue,
```
Comment (bdmogal) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/BusinessMetadataDataset.java]: this can be private

File: cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataHttpHandler.java (hunk 1 above)
Comment (henry-cask) [inline cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataHttpHandler.java]: Set since the returned data is the distinct objects inverted index by the value or key=value.
The dataset return list of actual records coming which could have duplicate target objects.

For example. If you search for value of "blah" and someone had 2 key-value pairs like key1=blah and key2=blah then the Dataset will return 2 records.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/service/BusinessMetadataStore.java (hunk 2 above)
Comment (henry-cask) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/service/BusinessMetadataStore.java]: Yeah, it will be decoded when come to server.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/BusinessMetadataDataset.java (hunk 3 above)
Comment (henry-cask) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/BusinessMetadataDataset.java]: Yes we could just use bytes instead.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/BusinessMetadataDatasetModule.java [hunk 6]
```diff
@@ -31,6 +31,6 @@ public void register(DatasetDefinitionRegistry registry) {
     DatasetDefinition<IndexedTable, ? extends DatasetAdmin> indexedTableDef =
       registry.get(IndexedTable.class.getName());
     registry.add(new BusinessMetadataDefinition("businessMetadataDataset", indexedTableDef));
-    registry.add(new BusinessMetadataDefinition(BusinessMetadataDataset.class.getSimpleName(), indexedTableDef));
+    registry.add(new BusinessMetadataDefinition(BusinessMetadataDataset.class.getName(), indexedTableDef));
```
Comment (bdmogal) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/BusinessMetadataDatasetModule.java]: why this change? is this a standard we use?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/service/BusinessMetadataStore.java (hunk 2 above)
Comment (bdmogal) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/service/BusinessMetadataStore.java]: :+1:

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/service/BusinessMetadataStore.java [hunk 7]
```diff
@@ -167,6 +169,34 @@ public Void apply(BusinessMdsIterable input) throws Exception {
...
+   */
+  public Iterable<BusinessMetadataRecord> searchMetadata(final String searchQuery) {
+    return searchMetadataOnType(searchQuery, MetadataSearchTargetType.ALL);
+  }
+
+  /**
+   * Search to the underlying Business Metadata Dataset for a target type.
+   */
+  public Iterable<BusinessMetadataRecord> searchMetadataOnType(final String searchQuery,
+                                                        final MetadataSearchTargetType type) {
```
Comment (bdmogal) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/service/BusinessMetadataStore.java]: nit: indentation

File: cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataHttpHandler.java (hunk 1 above)
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataHttpHandler.java]: Not a big deal, but we had used ``Set`` because the results from all these APIs are collections with unique elements.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/BusinessMetadataDatasetModule.java (hunk 6 above)
Comment (henry-cask) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/BusinessMetadataDatasetModule.java]: According to @poornachandra yes. We try to use getName instead.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/service/BusinessMetadataStore.java (hunk 7 above)
Comment (henry-cask) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/service/BusinessMetadataStore.java]: Got it. Thanks!

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/BusinessMetadataDataset.java (hunk 3 above)
Comment (henry-cask) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/BusinessMetadataDataset.java]: Actually lets keep it as String for now. It is used in the Row.getString and registration of BusinessMetadataDefinition. 
We could change it later.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/BusinessMetadataDataset.java (hunk 4 above)
Comment (henry-cask) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/BusinessMetadataDataset.java]: Will do.

File: cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataHttpHandler.java (hunk 1 above)
Comment (henry-cask) [inline cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataHttpHandler.java]: Sure. Set is fine too

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/BusinessMetadataDataset.java (hunk 3 above)
Comment (bdmogal) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/BusinessMetadataDataset.java]: hmmm, perhaps then create another constant - ``VALUE_COLUMN_BYTES`` and ``KEYVALUE_COLUMN_BYTES``, so we don't need conversion multiple times on every call?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/BusinessMetadataDataset.java (hunk 5 above)
Comment (henry-cask) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/BusinessMetadataDataset.java]: I set it as package private for now just in case we want to access it to do unit tests.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/service/BusinessMetadataStore.java (hunk 2 above)
Comment (henry-cask) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/service/BusinessMetadataStore.java]: But the design doc said we use key:value format so let me change it to that for now.

**Comment by henry-cask** [comment]
Succes build: 
http://builds.cask.co/browse/CDAP-DUT2812-2

File: cdap-app-fabric/src/main/java/co/cask/cdap/metadata/DefaultMetadataAdmin.java [hunk 8]
```diff
@@ -78,6 +85,33 @@ public void removeTags(Id.NamespacedId entityId, String... tags) throws NotFound
...
+    Set<MetadataSearchResultRecord> searchResultRecords = new LinkedHashSet<>();
+    Iterator<BusinessMetadataRecord> iterator = results.iterator();
+    while (iterator.hasNext()) {
+      BusinessMetadataRecord bmr = iterator.next();
+      MetadataSearchResultRecord msr =
+        new MetadataSearchResultRecord(bmr.getTargetId(),
+                                       (type == null ? MetadataSearchTargetType.ALL : type));
+
+      // Check if already added
+      if (!searchResultRecords.contains(msr)) {
```
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/metadata/DefaultMetadataAdmin.java]: since this is a set, this check is not required

File: cdap-app-fabric/src/main/java/co/cask/cdap/metadata/DefaultMetadataAdmin.java [hunk 9]
```diff
@@ -78,6 +85,33 @@ public void removeTags(Id.NamespacedId entityId, String... tags) throws NotFound
...
+    throws NotFoundException {
+    Iterable<BusinessMetadataRecord> results;
+    if (type == null) {
+      results = businessMds.searchMetadata(searchQuery);
+    } else {
+      results = businessMds.searchMetadataOnType(searchQuery, type);
+    }
+
+    Set<MetadataSearchResultRecord> searchResultRecords = new LinkedHashSet<>();
+    Iterator<BusinessMetadataRecord> iterator = results.iterator();
```
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/metadata/DefaultMetadataAdmin.java]: you should be able to use
```
for (BusinessMetadataRecord result : results) 
```
here?

**Comment by bdmogal** [comment]
@henry-cask just two minor comments. Not necessary to run build again. If you think they should be fixed, fix, squash and merge. LGTM

**Comment by henry-cask** [comment]
Thanks! Rebase, squash, and merging ...
